source/auth/auth.rst

```diff
@@ -1159,6 +1159,230 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
+the token(s) directly.
+
+The optional token refresh callback will accept the IDP information as
+well as the cached OIDCRequestTokenResult and return a new OIDCRequestTokenResult.
+
+
+.. code:: typescript
+
+    function onOIDCRefreshToken(serverInfo: OIDCMechanismServerStep1, tokenResult: OIDCRequestTokenResult): OIDCRequestTokenResult
+
+If the callback does not return an object in the correct form of ``OIDCRequestTokenResult``, the driver MUST raise an error.
+
+If the refresh callback is given and the request callback is not given,
+the driver MUST raise an error.  If DEVICE_NAME is given and one or more
+callbacks are given, the driver MUST raise an error.
+
+If no callbacks are given, the driver MUST enforce that a DEVICE_NAME
+mechanism_properties is set and one of ("aws", "azure", or "gcp").
+The callback mechanism can be used to support both Authentication
+Code Workflows or Device workflows that are not explicitly implemented
+by drivers.  If there is no callback and no DEVICE_NAME, or the
+DEVICE_NAME is set but credentials cannot be automatically obtained,
+the driver MUST raise an error.
+
+
+Supported Device Workflows
+``````````````````````````
+
+Drives MUST support device workflows for "aws", "azure", and "gcp", given
+by the DEVICE_NAME mechanism property.  In all cases the acquired token
+will be given as the ``jwt`` argument in the second client step of the
+SASL exchange.
+
+AWS
+___
+
+When the DEVICE_NAME mechanism property is set to "aws", the driver MUST
+attempt to read the value given by the ``AWS_WEB_IDENTITY_TOKEN_FILE`` and
+interpret it as a file path.  The contents of the file are read as the
+access token.
+
+Azure
+_____
+
+When the DEVICE_NAME mechanism property is set to "azure", the driver MUST
+acquire an access token using the workflow described in "Obtaining an Access Token for Azure Key Vault" <TODO insert link to client-side-encryption>.
+
+GCP
+___
+
+When the DEVICE_NAME mechanism property is set to "gcp", the driver MUST
+acquire an access token using the workflow described in "Obtaining GCP Credentials" <TODO insert link to client-side-encryption>.
+
+
+Caching Credentials
+```````````````````
+
+Drivers MUST enable caching when callback(s) are provided to the mongo client.
+When an authorization request is made and there is a valid cached response,
+the driver MUST use the cached response if it has not expired.
+
+The driver MAY, if it has previously acquired a valid Access Token, or has cached the reply from serverStep1, begin their authentication by sending clientStep2 with their initial saslStart request and authenticate in a single round trip.
+
+If the driver implements a global cache, the cache keys MUST include the
+principal name if given, and a hash of the callback function, if possible
+in the driver's language.  A global cache should be preferred, to prevent
+multiple browser interactions in the case of an authentication code
+workflow.
```

Comment:
Two things:

- In the devtools products, we won’t really be able to make use of this feature since we’ll need to maintain a cache across MongoClient instances in different processes. I could imagine that other products, if there are any, that use browser flows, would run into similar issues at some point. Consequently, I don’t know if adding global mutable state is worth it here – definitely not a blocker though!
- The cache key should probably include some form of cluster identifier, since the principal name may refer to different things for different clusters. Right now, on our side we’re planning to use the `client_id` provided in the initial SASL response for this.

Reply:
If we rely on a value from server step 1, we cannot skip client step 1 on subsequent calls.

Reply:
>The cache key should probably include some form of cluster identifier, since the principal name may refer to different things for different clusters. Right now, on our side we’re planning to use the client_id provided in the initial SASL response for this.

@addaleax , if we need to determine what cluster provided a cached value, is it easier to add a host name to a cache key instead tracking a client_id? It would be good to build a cache key based on initial client configuration and not on something dynamic. Also it will allow skipping step 1.

Reply:
Right, that’s a good point – I don’t mind using the host address as a key here if that’s easy enough. I’m not sure how common it would be to use the same client id for multiple clusters anyway.

Reply:
Updated